app/controllers/application_controller.rb

```diff
@@ -5,6 +5,151 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions 
+    set :session_secret, "password_security"
   end
 
+  get '/' do 
+    erb :index
+  end
+
+  get '/signup' do 
+    if !logged_in?
+      erb :'users/create_user'
+    else
+      redirect "/tweets"
+    end
+  end
+
+  post "/signup" do
+    user = User.new(username: params[:username], password: params[:password], password_confirmation: params[:password_confirmation], email: params[:email])
+    if params[:username].empty? || params[:password].empty? || params[:email].empty?
+      redirect "/signup", locals: {message: "Please use all three fields."}
+    else 
+      user.save
+      session["id"] = user.id
+      redirect "/tweets" 
+    end
+  end
+
+  get "/login" do
+    if !logged_in? 
+      erb :'users/login'
+    else 
+      redirect "/tweets"
+    end
+  end
+
+  post "/login" do
+    if params[:username].empty? || params[:password].empty?
+      redirect "/login"
+    else
+      user = User.find_by(username: params[:username])
+      if user && user.authenticate(params[:password])
+        session[:id] = user.id
+        redirect "/tweets" 
+      else
+        redirect "/"
+      end
+    end
+  end
+
+  get "/tweets" do
```

Comment:
This controller is overloaded. Consider separating your concerns by splitting some of these actions into separate controllers (i.e. a users controller and a tweets controller).